redis/lolwut.cc

```diff
@@ -0,0 +1,202 @@
+            px[j] = round(sin(k) * size + x);
+            py[j] = round(cos(k) * size + y);
+            k += M_PI/2;
+        }
+
+        for (int j = 0; j < 4; j++) {
+            draw_line(px[j], py[j], px[(j + 1) % 4], py[(j + 1) % 4], 1);
+        }
+    }
+
+    static canvas draw_schotter(const int cols, const int squares_per_row, const int squares_per_col) {
+        const int width = cols * 2;
+        const int padding = width > 4 ? 2 : 0;
+        const float side = static_cast<float>((width - padding * 1) / squares_per_row);
+        const int height = side * squares_per_col + padding * 0;
+        canvas c {width, height};
+        std::random_device seed_gen;
+        std::default_random_engine engine(seed_gen());
+        std::uniform_real_distribution<float> dist(0.0, 1.0);
+
+        for (int y = 0; y < squares_per_col; y++) {
+            for (int x = 0; x < squares_per_row; x++) {
+                int sx = x * side + side / 2 + padding;
+                int sy = y * side + side / 2 + padding;
+
+                float angle = 0;
+                if (y > 1) {
+                    float r1 = dist(engine) / squares_per_col * y;
+                    float r2 = dist(engine) / squares_per_col * y;
+                    float r3 = dist(engine) / squares_per_col * y;
+                    if (engine() % 2) {
+                        r1 = -r1;
+                    }
+                    if (engine() % 2) {
+                        r2 = -r2;
+                    }
+                    if (engine() % 2) {
+                        r3 = -r3;
+                    }
+                    angle = r1;
+                    sx += r2 * side / 3;
+                    sy += r3 * side / 3;
+                }
+                c.draw_square(sx, sy, side, angle);
+            }
+        }
+        return c;
+    }
+
+    bytes render_canvas() {
+        bytes txt;
+        for (int y = 0; y < _height; y += 4) {
+            for (int x = 0; x < _width; x += 2) {
+                int byte = 0;
+                int8_t unicode[3];
+                if (get_pixel(x, y)) {
+                    byte |= (1 << 0);
+                }
+                if (get_pixel(x, y + 1)) {
+                    byte |= (1 << 1);
+                }
+                if (get_pixel(x, y + 2)) {
+                    byte |= (1 << 2);
+                }
+                if (get_pixel(x + 1, y)) {
+                    byte |= (1 << 3);
+                }
+                if (get_pixel(x + 1, y + 1)) {
+                    byte |= (1 << 4);
+                }
+                if (get_pixel(x + 1, y + 2)) {
+                    byte |= (1 << 5);
+                }
+                if (get_pixel(x, y + 3)) {
+                    byte |= (1 << 6);
+                }
+                if (get_pixel(x + 1, y + 3)) {
+                    byte |= (1 << 7);
+                }
+                translate_pixels_group(byte, unicode);
+                txt.append(unicode, 3);
```

Comment:
This has quadratic behavior. Better to allocate txt up-front and copy to the correct location.

Reply:
Another option is to accept a bytes_ostream parameter which knows how to grow without quadratic behavior.